File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {        
        vector<vector<int>> result;
        if (root == NULL) return result;
        
        queue<TreeNode *> Q;
        Q.push(root);
                    
        while (!Q.empty()){
            int n = Q.size();  // # of nodes at this level
            
            // Add a new list to the end of the result
            result.push_back({});
            int index = result.size()-1;
            
            for (int i=0; i<n; i++){
                auto p = Q.front();
                Q.pop();
                result[index].push_back(p->val);
                if (p->left) Q.push(p->left);
                if (p->right) Q.push(p->right);                    
            } // end-for
        } //end-while        

        return result;
    } // end-levelOrder
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if root is None: return result

        Q = []
        Q.append(root)
        
        while len(Q) > 0:
            n = len(Q)  # Number of nodes at this level
            
            # Add a new list to the end of the result
            result.append([])
            index = len(result)-1

            for i in range(n):
                p = Q.pop(0)
                result[index].append(p.val)
                if p.left is not None: Q.append(p.left)
                if p.right is not None: Q.append(p.right)
        return result